ingenialink/ethercat/network.py

```diff
@@ -62,52 +66,96 @@ def __init__(self, interface_name):
-        """Loads a given firmware file to a target.
-
-        .. warning::
-            Choose the ``boot_in_app`` flag accordingly to your
-            servo specifications otherwise the servo could enter
-            a blocking state.
+    def load_firmware(self, fw_file, slave_id=1, boot_in_app=None):
+        """Loads a given firmware file to a target slave.
 
         .. warning ::
             It is needed to disconnect the drive(:func:`disconnect_from_slave`)
             after loading the firmware since the `Servo` object's data will
             become obsolete.
 
         Args:
-            target (int): Targeted node ID to be loaded.
             fw_file (str): Path to the firmware file.
+            slave_id (int): Slave ID to which load the firmware file.
             boot_in_app (bool): If ``fw_file`` extension is .sfu -> True.
                                 Otherwise -> False.
 
         Raises:
-            ILFirmwareLoadError: The firmware load process fails
-                with an error message.
+            FileNotFoundError: If the firmware file cannot be found.
             ValueError: If the firmware file has the wrong extension.
+            ILFirmwareLoadError: If no slave is detected.
+            ILFirmwareLoadError: If the FOE write operation is not
+            successful.
 
         """
         if not os.path.isfile(fw_file):
-            raise FileNotFoundError("Could not find {}.".format(fw_file))
-
+            raise FileNotFoundError(f"Could not find {fw_file}.")
         if boot_in_app is None:
             if not fw_file.endswith((FILE_EXT_SFU, FILE_EXT_LFU)):
-                raise ValueError(
-                    f"Firmware file should have extension {FILE_EXT_SFU} or {FILE_EXT_LFU}"
-                )
+                raise ValueError("Firmware file should have extension "
+                                 f"{FILE_EXT_SFU} or {FILE_EXT_LFU}.")
             boot_in_app = fw_file.endswith(FILE_EXT_SFU)
-        self._cffi_network = ffi.new("il_net_t **")
-        _interface_name = cstr(self.interface_name) if self.interface_name else ffi.NULL
-        _fw_file = cstr(fw_file) if fw_file else ffi.NULL
-        r = lib.il_net_update_firmware(
-            self._cffi_network, _interface_name, target, _fw_file, boot_in_app
-        )
+        self._ecat_master = Master()
+        self._ecat_master.open(self.interface_name)
+        if self._ecat_master.config_init() < 0:
+            raise ILFirmwareLoadError("Firmware could not be loaded. "
+                                      "No slave detected.")
+        slave = self._ecat_master.slaves[slave_id-1]
+        if not boot_in_app:
+            slave.sdo_write(FORCE_COCO_BOOT_IDX,
+                            FORCE_COCO_BOOT_SUBIDX,
+                            FORCE_BOOT_PASSWORD.to_bytes(4, 'little'),
+                            False)
+            # COMOCO drives need to be forced two times into bootstrap state
```

Comment:
Could you let me know if that was discussed with the FW team?

Reply:
I did not discuss this with the FW team. I followed [this ](https://drives.novantamotion.com/summit/foe-file-over-ethercat-bootloader) document.